### PR TITLE
openocd: Check for a connection several times

### PR DESCRIPTION
**Pull request by dcz-self**
OpenOCD can start slowly under system load. Instead of trying once and giving up, this tries a couple times for a few seconds. In the best case, there's no extra time spent waiting. In the worst case, we can still connect to a delayed OpenOCD.

It's the last fix to OpenOCD that I had waiting :)

### CODE REVIEW

File: tockloader/openocd.py [hunk 1]
```diff
@@ -408,16 +408,23 @@ def run_terminal(self):
...
             listener = socket.socket()
-            listener.connect(("127.0.0.1", 9999))
+            for i in reversed(range(3)):
+                # Delay to give the connection time to start before running
+                # the RTT listener.
+                time.sleep(1)
+                if ocd_p.poll():
+                    return
+                try:
+                    listener.connect(("127.0.0.1", 9999))
```
Comment (hudson-ayers) [inline tockloader/openocd.py:419]:
```suggestion
                    logging.status("Listening for messages, attempt {}", i+1);
                    listener.connect(("127.0.0.1", 9999))
```

File: tockloader/openocd.py [hunk 2]
```diff
@@ -408,16 +408,23 @@ def run_terminal(self):
...
-            # Delay to give the connection time to start before running
-            # the RTT listener.
-            time.sleep(1)
-            if ocd_p.poll():
-                return
-
-            logging.status("Listening for messages.")
             listener = socket.socket()
-            listener.connect(("127.0.0.1", 9999))
+            for i in reversed(range(3)):
```
Comment (hudson-ayers) [inline tockloader/openocd.py:412]:
```suggestion
            const MAX_TRIES = 3;
            for i in 0..MAX_TRIES:
```

This is just personal preference, but I think it is a bit less confusing

File: tockloader/openocd.py [hunk 3]
```diff
@@ -408,16 +408,23 @@ def run_terminal(self):
...
+                # Delay to give the connection time to start before running
+                # the RTT listener.
+                time.sleep(1)
+                if ocd_p.poll():
+                    return
+                try:
+                    listener.connect(("127.0.0.1", 9999))
+                    break
+                except ConnectionRefusedError:
+                    if i == 0:
```
Comment (hudson-ayers) [inline tockloader/openocd.py:422]:
```suggestion
                    if i == MAX_TRIES - 1:
```

Comment (dcz-self) [reply tockloader/openocd.py]: I don't like the off-by-one errors that this invites (I actually tried to recreate this from memory and blundered in https://github.com/tock/tockloader/commit/16c2a2d00f349eba6b0063e0dd6f845a5f03bad5#diff-387adfb447164526e8d43f186bf54c3532fb67df42a623a2d2b829e3628fe6dbR424 ), but if this gets the change upstreamed, it's good enough.

Comment (hudson-ayers) [reply tockloader/openocd.py]: I just realized I gave a bunch of Rust syntax suggestions in a python file

Comment (dcz-self) [reply tockloader/openocd.py]: No worries, I already pushed an updated version.

**Review by hudson-ayers — APPROVED** [review]

**Review by bradjc — APPROVED** [review]

**Pull request merged**